src/disk_usage.rs

```diff
@@ -0,0 +1,253 @@
+/// potential increase in storage from compression, as Snappy does not guarantee
+/// compression results in fewer bytes.
+const EXTRA_PADDING: u64 = 1024 * 100;
+
+#[derive(Debug)]
+struct DiskUsage {
+    pub total: u64,
+    pub used: u64,
+}
+
+/// Check an *estimate* of the disk usage of a snapshot
+///
+/// This is a best effort attempt to estimate the disk usage of a snapshot and
+/// validate the snapshot will fit in the configured parameters.
+pub(crate) fn check(
+    image_path: &Path,
+    memory_ranges: &[Range<u64>],
+    max_disk_usage: Option<NonZeroU64>,
+    max_disk_usage_percentage: Option<f64>,
+) -> Result<()> {
+    let estimate_add = estimate(memory_ranges);
+
+    if let Some(max_disk_usage) = max_disk_usage {
+        check_max_usage(estimate_add, max_disk_usage)?;
+    }
+
+    if let Some(max_disk_usage_percentage) = max_disk_usage_percentage {
+        let disk_usage = disk_usage(image_path)?;
+        check_max_usage_percentage(estimate_add, &disk_usage, max_disk_usage_percentage)?;
+    }
+
+    Ok(())
+}
+
+fn check_max_usage(estimate_add: u64, max_disk_usage: NonZeroU64) -> Result<()> {
+    // convert to MB
+    let max_disk_usage_mb = max_disk_usage.get() * 1024 * 1024;
+
+    if estimate_add > max_disk_usage_mb {
+        return Err(Error::DiskUsageEstimateExceeded(
+            estimate_add,
+            max_disk_usage_mb,
+        ));
+    }
+    Ok(())
+}
+
+fn check_max_usage_percentage(
+    estimate_add: u64,
+    disk_usage: &DiskUsage,
+    max_disk_usage_percentage: f64,
+) -> Result<()> {
+    let estimated_used = disk_usage.used.saturating_add(estimate_add);
+
+    // assuming the disk was empty, how much could we use
+    let max_allowed =
+        f64_to_u64(u64_to_f64(disk_usage.total)? * (max_disk_usage_percentage / 100.0))?;
+
+    if estimated_used > max_allowed {
+        let allowed = max_allowed.saturating_sub(disk_usage.used);
+        return Err(Error::DiskUsageEstimateExceeded(estimate_add, allowed));
+    }
+
+    Ok(())
+}
+
+/// Attempt to convert u64 into f64
+///
+/// Of note: The maximum value works out to be more than 18 exabytes, which is
+/// *way* more than any memory or disk we're likely to see any time soon.
+///
+/// `TryInto<f64> for u64` is not implemented.  This tries to be mindful of the
+/// following edge condition:
+/// 1. The value must be less than `f64::MAX - 1.0`
+#[allow(clippy::cast_precision_loss)]
+fn u64_to_f64(value: u64) -> Result<f64> {
+    if value > f64_to_u64(f64::MAX - 1.0)? {
+        return Err(Error::Other(
+            "unable to convert u64 to f64",
+            format!("value is too large to convert to f64: {value}"),
+        ));
+    }
+    Ok(value as f64)
+}
+
+/// Attempt to convert f64 into u64
+///
+/// `TryInto<u64> for f64` is not implemented.  This tries to be mindful of the
+/// following edge conditions:
+/// 1. The value must be a signed positive value
+/// 2. The value is explicitly truncated to the integer value
```

Comment:
nit: Consider saying "truncated and clamped". `f64::MAX` is many orders of magnitude greater than `u64::MAX`, for  instance, and `as u64` clamps it to `u64::MAX`.